packages/manager/apps/pci-users/src/components/users/EditRolesModal.tsx

```diff
@@ -0,0 +1,144 @@
+import {
+  OsdsButton,
+  OsdsCheckbox,
+  OsdsCheckboxButton,
+  OsdsModal,
+  OsdsSpinner,
+  OsdsText,
+} from '@ovhcloud/ods-components/react';
+import {
+  ODS_BUTTON_VARIANT,
+  ODS_CHECKBOX_BUTTON_SIZE,
+  ODS_SPINNER_SIZE,
+  OdsCheckboxCheckedChangeEventDetail,
+  OsdsCheckboxCustomEvent,
+} from '@ovhcloud/ods-components';
+import {
+  ODS_THEME_COLOR_INTENT,
+  ODS_THEME_TYPOGRAPHY_LEVEL,
+} from '@ovhcloud/ods-common-theming';
+import { useTranslation } from 'react-i18next';
+import { useEffect, useState } from 'react';
+import { useUserRoles } from '@/hooks/useUser';
+import { useAllRoles, useUpdateUserRoles } from '@/hooks/useRole';
+
+interface EditRolesProps {
+  projectId: string;
+  userId: number;
+  onClose: () => void;
+  onSuccess: () => void;
+  onError: (cause: Error) => void;
+}
+
+export default function EditRolesModal({
+  projectId,
+  userId,
+  onClose,
+  onSuccess,
+  onError,
+}: EditRolesProps) {
+  const { t } = useTranslation('common');
+  const { data: roles, isPending: isPendingRoles } = useAllRoles(
+    `${projectId}`,
+  );
+  const { data: userRoles, isPending } = useUserRoles(
+    `${projectId}`,
+    `${userId}`,
+  );
+  const { isPending: isPendingUpdate, update } = useUpdateUserRoles({
+    projectId: `${projectId}`,
+    userId,
+    onError: (err) => {
+      onClose();
+      onError(err);
+    },
+    onSuccess: () => {
+      onClose();
+      onSuccess();
+    },
+  });
+  const [userRoleId, setUserRoleId] = useState<string[]>([]);
+  useEffect(() => {
+    if (userRoles) {
+      setUserRoleId(
+        userRoles.map((role) => {
+          return role.id;
+        }),
+      );
+    }
+  }, [userRoles]);
+  return (
+    <>
+      <OsdsModal
+        headline={t('pci_projects_project_users_roles')}
+        onOdsModalClose={onClose}
+      >
+        <slot name="content">
+          {!isPending && !isPendingRoles && !isPendingUpdate && (
```

Comment:
This result is used in three different places, you can move it to a useMemo and do a ternary here